orchestrator/utils/search_query.py

```diff
@@ -0,0 +1,348 @@
+        TSQueryVisitor.visit_term(key_node, acc)
+        acc.append(" <-> ")
+
+        if value_node[0] == "ValueGroup":
+            acc.append("(")
+            for v in value_node[1]:
+                TSQueryVisitor.visit_search_word(v, acc)
+                acc.append(" | ")
+            acc.pop()
+            acc.append(")")
+        else:
+            TSQueryVisitor.visit_term(value_node, acc)
+
+    @staticmethod
+    def visit_search_word(node: Node, acc: list[str]) -> None:
+        # Postgres is finicky with single quotes in the query. In some situations it throws an error.
+        # To avoid the special handling of ' by PG, we replace it with double quotes.
+        text = node[1].replace("'", '"')
+        if node[0] == "Word":
+            acc.append(text)
+        elif node[0] == "PrefixWord":
+            acc.append(f"{text}:*")
+        else:
+            raise Exception(f"Invalid SearchWord type {node[0]}")
+
+    @staticmethod
+    def visit_phrase(node: Node, acc: list[str]) -> None:
+        for search_word in node[1]:
+            TSQueryVisitor.visit_search_word(search_word, acc)
+            acc.append(" <-> ")
+
+        if len(node[1]):
+            acc.pop()  # Remove trailing space
+
+    @staticmethod
+    def visit_term(node: Node, acc: list[str]) -> None:
+        node_type = node[0]
+        if node_type == "Phrase":
+            TSQueryVisitor.visit_phrase(node, acc)
+        elif node_type in ["Word", "PrefixWord"]:
+            TSQueryVisitor.visit_search_word(node, acc)
+        elif node_type == "KVTerm":
+            TSQueryVisitor.visit_kv_term(node, acc)
+        elif node_type == "Group":
+            TSQueryVisitor.visit_group(node, acc)
+        elif node_type == "Negation":
+            acc.append("!")
+            should_group = node[1][0] in ["Phrase", "KVTerm"]
+            if should_group:
+                acc.append("(")
+            TSQueryVisitor.visit_term(node[1], acc)
+            if should_group:
+                acc.append(")")
+        else:
+            raise Exception(f"Unexpected term node type: {node_type}")
+
+    @staticmethod
+    def visit_and_expression(node: Node, acc: list[str]) -> None:
+        for term in node[1]:
+            TSQueryVisitor.visit_term(term, acc)
+            acc.append(" & ")
+        if len(node[1]):
+            acc.pop()
+
+    @staticmethod
+    def visit_query(node: Node, acc: list[str]) -> None:
+        for expression in node[1]:
+            TSQueryVisitor.visit_and_expression(expression, acc)
+            acc.append(" | ")
+
+        if len(node[1]):
+            acc.pop()
+
+    @staticmethod
+    def visit(parse_tree: Node) -> str:
+        acc: list[str] = []
+        node_type = parse_tree[0]
+        if node_type == "Query":
+            TSQueryVisitor.visit_query(parse_tree, acc)
+        return "".join(acc)
+
+
+def create_ts_query_string(search_query: str) -> str:
+    return TSQueryVisitor.visit(Parser(Lexer(search_query).lex()).parse())
```

Comment:
I think it would be good to define an upper limit for the query. Maybe based on the number of nodes?

Reply:
You mean an upper limit to the size of the query?

We would have to track the nesting level. Maybe easier to limit the number of characters in the submitted query string?

Reply:
I think something more detailed than query size is needed to stop queries `" ".join("a" for i in range(10_000))` etc. On my machine this takes 2-3 seconds to complete, and the final PG query might take even longer.

With the target users of this library there might not be any concerns of deliberate abuse but we might as well prevent it.

Maybe we can add such a constraint to the accumulator, make it wrap a list and and raise an error when appending/extending it beyond a certain size?

Reply:
That seems short, but sure let's start there :) If it is an ENV setting we can change it adhoc if need be.

Reply:
I made it 512 :laughing: